### PR TITLE
[memory][context] Local user permissions for complete actions in sc-structures of knowledge base

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

### CODE REVIEW

File: sc-memory/sc-memory/sc_memory.cpp [hunk 1]
```diff
@@ -827,46 +827,9 @@ bool ScMemoryContext::HelperResolveSystemIdtf(
...
-        "Not able to set system identifier due sc-memory context hasn't erase access levels");
-
-  default:
-    break;
-  }
-
-  return result == SC_RESULT_OK;
+  ScSystemIdentifierQuintuple quintuple;
+  return HelperSetSystemIdtf(sysIdtf, addr, quintuple);
+  ;
```
Comment (MksmOrlov) [inline sc-memory/sc-memory/sc_memory.cpp:838]: unused ;

File: sc-memory/sc-memory/sc_memory.cpp [hunk 2]
```diff
@@ -827,46 +827,9 @@ bool ScMemoryContext::HelperResolveSystemIdtf(
...
-        utils::ExceptionInvalidState,
-        "Not able to set system identifier due sc-memory context hasn't erase access levels");
-
-  default:
-    break;
-  }
-
-  return result == SC_RESULT_OK;
+  ScSystemIdentifierQuintuple quintuple;
+  return HelperSetSystemIdtf(sysIdtf, addr, quintuple);
```
Comment (MksmOrlov) [inline sc-memory/sc-memory/sc_memory.cpp:837]: verify system idtf not to duplicate

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Review by KovalM — CHANGES_REQUESTED** [review]

File: sc-memory/sc-core/sc-store/sc_iterator3.c [hunk 3]
```diff
@@ -204,6 +204,8 @@ sc_iterator3 * sc_iterator3_new(
 
   it->type = type;
   it->ctx = ctx;
+  it->search_structure = SC_FALSE;
+  it->global_search = SC_FALSE;
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_iterator3.c]: It is not actual

File: sc-memory/sc-core/sc-store/sc_iterator3.h [hunk 4]
```diff
@@ -47,6 +47,8 @@ struct _sc_iterator3
   sc_iterator_param params[3];    // parameters array
   sc_addr results[3];             // results array (same size as params)
   sc_memory_context const * ctx;  // pointer to used memory context
+  sc_bool search_structure;
+  sc_bool global_search;
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_iterator3.h]: It is not actual

File: sc-memory/sc-core/sc_memory.c [hunk 5]
```diff
@@ -226,7 +224,8 @@ sc_uint32 sc_memory_get_element_output_arcs_count(sc_memory_context const * ctx,
     return 0;
   }
 
-  if (_sc_memory_context_check_global_access_levels(memory->context_manager, ctx, SC_CONTEXT_ACCESS_LEVEL_READ)
+  if (_sc_memory_context_check_local_and_global_access_levels(
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory.c:227]: I don't remember this. Let's discuss

File: sc-memory/sc-core/sc_memory.c [hunk 6]
```diff
@@ -338,11 +325,25 @@ sc_addr sc_memory_arc_new_ext(sc_memory_context const * ctx, sc_type type, sc_ad
     return SC_ADDR_EMPTY;
   }
 
-  if (_sc_memory_context_check_global_access_levels(memory->context_manager, ctx, SC_CONTEXT_ACCESS_LEVEL_WRITE)
-      == SC_FALSE)
+  if (_sc_memory_context_check_if_is_accessed_structure(
+          memory->context_manager, ctx, SC_CONTEXT_ACCESS_LEVEL_WRITE, beg)
+          == SC_FALSE
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory.c:330]: We check that the provided begin sc-address is an sc-address of sc-structure that differs user access levels for sc-constructions in this sc-structure. 
1) If provided begin sc-address is an sc-address of such sc-structure and we want to generate access arc from it to some sc-element then we don't check the end sc-address. It is the case when user add something in your own sc-structure. He can do it if he has access levels for writing in this sc-structure.
2)  If provided begin sc-address is an sc-address of not such sc-structure then we check end sc-address.

**Review by MksmOrlov — APPROVED** [review]

File: sc-memory/sc-core/sc_memory.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-core/sc_memory.c:226]: According our discuss, use 'permission' term

**Review by KovalM — CHANGES_REQUESTED** [review]
Use 'permission' term for access_levels

**Review by SemSystemsInternal — APPROVED** [review]

**Pull request merged**